### PR TITLE
I added a DB, but I am afraid that my app was broked=(

### PR DESCRIPTION
**Pull request by VasylKits**

### CODE REVIEW

**Comment by VasylKits** [comment]
Done, but I think that all method must have to be Async?

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 1]
```diff
@@ -11,81 +14,108 @@ namespace AnnouncementWebApi.Controllers
...
-            announcements.Add(newAnn);
+            _myDbContext.Announcements.Add(newAnn);
+
+            try
+            {
+                _myDbContext.SaveChanges();
+            }
+            catch
+            {
+                throw new Exception();
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:61]: Show exception message to the user.

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 2]
```diff
@@ -11,81 +14,108 @@ namespace AnnouncementWebApi.Controllers
...
         [HttpPut]
-        public IActionResult EditAnnouncement([FromBody]EditAnnouncement editAnnouncement)
+        public IActionResult EditAnnouncement([FromBody] EditAnnouncement editAnnouncement)
         {
-            var editAnn = announcements.SingleOrDefault(a => a.Id == editAnnouncement.Id);
+            var editAnn = _myDbContext.Announcements.SingleOrDefault(a => a.Id == editAnnouncement.Id);
             if (editAnn == null)
             {
-                return NotFound();
+                return BadRequest();
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:73]: Show the message to the user.

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 3]
```diff
@@ -11,81 +14,108 @@ namespace AnnouncementWebApi.Controllers
...
                 editAnn.Title = editAnnouncement.Title;
                 editAnn.Description = editAnnouncement.Description;
                 editAnn.EditDate = DateTime.Now;
-                return Ok(editAnn);
+                _myDbContext.SaveChanges();
             }
-//            return Ok(editAnn);
+            catch (DbUpdateConcurrencyException)
+            {
+                throw;
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:87]: Show exception message to the user.

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 4]
```diff
@@ -11,81 +14,108 @@ namespace AnnouncementWebApi.Controllers
...
-            if (announcement == null)
+            try
             {
-                return NotFound();
+                _myDbContext.SaveChanges();
             }
-            return Ok(announcement);
+            catch
+            {
+                throw new Exception();
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:110]: Show exception message to the user.

File: AnnouncementWebApi/Startup.cs [hunk 5]
```diff
@@ -19,6 +17,26 @@ public class Startup
         public Startup(IConfiguration configuration)
         {
             Configuration = configuration;
+            string dbName = "TestDatabase.db";
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Startup.cs:20]: Change DB name to an understandable name.

File: AnnouncementWebApi/Startup.cs [hunk 6]
```diff
@@ -19,6 +17,26 @@ public class Startup
...
+            if (File.Exists(dbName))
+            {
+                File.Delete(dbName);
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Startup.cs:23]: Why do we need this deletion every time when we are running the project?

File: AnnouncementWebApi/DB/DbContext.cs [hunk 7]
```diff
@@ -1,6 +1,7 @@
-﻿namespace AnnouncementWebApi.DB
+﻿using Microsoft.EntityFrameworkCore;
+namespace AnnouncementWebApi.DB
 {
-    public class DbContext
-    {
-    }
+    //public class DbContext
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/DB/DbContext.cs:4]: Delete the class

File: AnnouncementWebApi/DB/MyDbContext.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
+﻿using Microsoft.Data.Sqlite;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Sqlite;
+using System.IO;
+using System.Reflection;
+
+namespace AnnouncementWebApi.DB
+{
+    public class MyDbContext : DbContext
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/DB/MyDbContext.cs:9]: Rename the class to an understandable name.

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 9]
```diff
@@ -11,81 +14,108 @@ namespace AnnouncementWebApi.Controllers
...
+            try
             {
                 editAnn.Title = editAnnouncement.Title;
                 editAnn.Description = editAnnouncement.Description;
                 editAnn.EditDate = DateTime.Now;
-                return Ok(editAnn);
+                _myDbContext.SaveChanges();
             }
-//            return Ok(editAnn);
+            catch (DbUpdateConcurrencyException)
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:85]: Remove redundant catch statement or change the exception type.

File: AnnouncementWebApi/Startup.cs [hunk 10]
```diff
@@ -19,6 +17,26 @@ public class Startup
...
+            }
+            using (var dbContext = new MyDbContext())
+            {
+                dbContext.Database.EnsureCreated();
+                if (!dbContext.Announcements.Any())
+                {
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Startup.cs:29]: We don't need this pre-initialization.

**Review by nikoliasoroka — COMMENTED** [review]
1. Delete unnecessary comments;
2. Delete unnecessary usings;
3. Change controller's actions to Async versions;
4. Why do we need Id in the NewAnnouncement class?
5. What is the goal of DelAnnouncement class?
6. Move NewAnnouncement and EditAnnouncement to the DTOa folder;

File: AnnouncementWebApi/Startup.cs (hunk 6 above)
Comment (VasylKits) [reply AnnouncementWebApi/Startup.cs:23]: If I delete this operation than all added and edit announcements are saved in database.

File: AnnouncementWebApi/DB/DbContext.cs (hunk 7 above)
Comment (VasylKits) [reply AnnouncementWebApi/DB/DbContext.cs:4]: +

File: AnnouncementWebApi/DB/MyDbContext.cs (hunk 8 above)
Comment (VasylKits) [reply AnnouncementWebApi/DB/MyDbContext.cs:9]: DatabaseContext

File: AnnouncementWebApi/Controllers/AnnouncementController.cs (hunk 9 above)
Comment (VasylKits) [reply AnnouncementWebApi/Controllers/AnnouncementController.cs:85]: `catch (Exception ex)
            {
                throw new Exception(ex.Message);
            }
            return Ok(editAnnouncement);`

File: AnnouncementWebApi/Startup.cs (hunk 10 above)
Comment (VasylKits) [reply AnnouncementWebApi/Startup.cs:29]: I have created a null datadase?

**Comment by VasylKits** [comment]
I don't know how wroting the code for creating database =(

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 11]
```diff
@@ -47,75 +33,62 @@ public IActionResult GetById(int id)
...
             try
             {
-                _myDbContext.SaveChanges();
+                await _myDbContext.SaveChangesAsync();
             }
-            catch
+            catch (Exception ex)
             {
-                throw new Exception();
+                throw new Exception(ex.Message);
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:46]: Use the appropriate method to return the exception message for the user.

File: AnnouncementWebApi/Controllers/AnnouncementController.cs [hunk 12]
```diff
@@ -47,75 +33,62 @@ public IActionResult GetById(int id)
...
             }
-            catch (DbUpdateConcurrencyException)
+            try
+            {
+                await _myDbContext.SaveChangesAsync();
+            }
+            catch (Exception ex)
             {
-                throw;
+                throw new Exception(ex.Message);
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Controllers/AnnouncementController.cs:70]: Use the appropriate method to return the exception message for the user.

File: AnnouncementWebApi/DTOs/NewAnnouncement.cs [hunk 13]
```diff
@@ -1,11 +1,7 @@
-﻿using System;
-using System.Collections.Generic;
-
-namespace AnnouncementWebApi
+﻿namespace AnnouncementWebApi
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/DTOs/NewAnnouncement.cs:1]: Incorrect namespace!

File: AnnouncementWebApi/Models/Announcement.cs [hunk 14]
```diff
@@ -1,6 +1,4 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.ComponentModel.DataAnnotations;
 
 namespace AnnouncementWebApi
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Models/Announcement.cs:3]: Incorrect namespace!

Comment (nikoliasoroka) [reply AnnouncementWebApi/Models/Announcement.cs:3]: Are you sure we need this: EditData = null?
What is the default value for nullable types?

File: AnnouncementWebApi/Startup.cs (hunk 10 above)
Comment (nikoliasoroka) [reply AnnouncementWebApi/Startup.cs:23]: You should create an empty DB.

**Review by nikoliasoroka — COMMENTED** [review]
Go through the comments

File: AnnouncementWebApi/Startup.cs [hunk 15]
```diff
@@ -19,6 +15,9 @@ public class Startup
         public Startup(IConfiguration configuration)
         {
             Configuration = configuration;
+            using var dbContext = new DatabaseContext();
+            dbContext.Database.EnsureCreated();
+            dbContext.Announcements = null;
```
Comment (nikoliasoroka) [inline AnnouncementWebApi/Startup.cs:20]: Why do we need it? I think it will work without a null assignment.

**Comment by VasylKits** [comment]
Add BadRequest and modified database

**Review by nikoliasoroka — APPROVED** [review]

**Pull request merged**